File: dashboard/static/js/standings.js
/* The standings: every account on this exchange, ranked by what it made.
 *
 * The venue ships with a population of agents, and it also lets outside
 * systems in through the public API. The question this screen exists to answer
 * is whether those outside systems beat the agents the market ships with, so
 * residents and seats are ranked in one list against each other, and a seat is
 * marked to be findable at a glance down a column of two dozen agents.
 *
 * The venue's own accounts are listed apart and are not ranked at all. A fee
 * treasury does have a return and the arithmetic behind it is sound, but it
 * measures fees taken against fees taken rather than a trader against a
 * market. In the same list it would win, which is the wrong answer to the only
 * question the screen is asking.
 */

import { clock, cls, count, esc, money, signed } from './format.js';

/** The mark `format.js` prints where a number is absent. One glyph, one meaning. */
const ABSENT = '-';

/* `clock` reads the nanoseconds the tape is stamped in. `as_of` counts in
 * simulated seconds, the same clock the lab screen reports as uptime. */
const NS_PER_SECOND = 1e9;

const KIND_WORDS = { resident: 'Resident', seat: 'Seat', operator: 'Venue' };

/**
 * The number a field holds, or null where it holds none.
 *
 * `Number(null)` and `Number('')` are both 0, and 0 is finite, so every absent
 * field on this payload would otherwise sort, colour and print as a flat zero.
 * An account that is flat and an account whose equity the venue could not
 * compute are different facts, and a leaderboard is exactly the table that
 * would quietly confuse them.
 */
function num(value) {
  if (value == null || value === '') return null;
  const n = Number(value);
  return Number.isFinite(n) ? n : null;
}

/**
 * A return, as a percentage.
 *
 * `return_pct` arrives already in percent units, so `percent()` is the wrong
 * helper here: it multiplies by a hundred and would report 4.555 as 455.5%.
 * `signed` gives a gain the plus a leaderboard needs and `money` does not
 * supply, and refuses anything it cannot make a number of.
 */
function ret(value) {
  const text = signed(value, 2);
  return text === ABSENT ? ABSENT : `${text}%`;
}

/** Money, carrying the sign a gain does not otherwise get. */
function gain(value) {
  const text = money(value);
  if (text === ABSENT) return ABSENT;
  return (num(value) ?? 0) > 0 ? `+${text}` : text;
}

function whole(value) {
  const n = num(value);
  return n == null ? ABSENT : count(n);
}

/**
 * Best first, and unknown last.
 *
 * The venue ranks, and its rank is the authority wherever it gave one. What
 * this settles is the rest: a row the venue left unranked, or whose return it
 * does not know, cannot be placed against one it does, so it goes to the end
 * rather than to the top, where a missing number read as a zero would put an
 * unmeasured account above everything that lost money.
 */
function byStanding(a, b) {
  const ra = num(a.rank);
  const rb = num(b.rank);
  if (ra != null && rb != null) return ra - rb;
  if (ra != null || rb != null) return ra != null ? -1 : 1;
  const pa = num(a.return_pct);
  const pb = num(b.return_pct);
  if (pa != null && pb != null) return pb - pa;
  if (pa != null || pb != null) return pa != null ? -1 : 1;
  return 0;
}

/**
 * The kind, as the one thing on the row a reader is scanning for.
 *
 * Amber is the chrome colour on this terminal and is what carries emphasis;
 * green and red are spoken for by direction and may not be borrowed to mean a
 * category. Only the seat is badged, because a badge on all three kinds is a
 * uniform, and a uniform distinguishes nobody.
 */
function tag(kind) {
  const word = KIND_WORDS[kind] ?? (kind == null ? ABSENT : String(kind));
  if (kind === 'seat') return `<span class="badge amber">${esc(word)}</span>`;
  return `<span class="faint">${esc(word)}</span>`;
}

/* A halted account is a fact about this session; a quiet one is not. Only
 * `true` earns the badge, so a venue that says nothing makes no claim. */
function who(entry) {
  const label = String(entry.label ?? entry.id ?? '');
  const id = String(entry.id ?? '');
  return `<td style="text-align:left">
    <b>${esc(label)}</b>
    ${id && id !== label ? `<span class="faint"> ${esc(id)}</span>` : ''}
    ${entry.halted === true ? ' <span class="badge closed">Halted</span>' : ''}
  </td>`;
}

function traderRow(entry) {
  return `<tr data-kind="${esc(entry.kind ?? 'unknown')}">
    <td style="text-align:right" class="faint">${whole(entry.rank)}</td>
    ${who(entry)}
    <td style="text-align:left">${tag(entry.kind)}</td>
    <td class="${cls(num(entry.return_pct))}">${ret(entry.return_pct)}</td>
    <td class="${cls(num(entry.pnl))}">${gain(entry.pnl)}</td>
    <td>${money(entry.equity)}</td>
    <td class="faint">${whole(entry.positions)}</td>
  </tr>`;
}

/* No rank and no return, which is the whole point of the separation: both
 * compute for a fee account, and neither says anything true about it. */
function venueRow(entry) {
  return `<tr data-kind="${esc(entry.kind ?? 'operator')}">
    ${who(entry)}
    <td>${money(entry.equity)}</td>
    <td class="${cls(num(entry.pnl))}">${gain(entry.pnl)}</td>
    <td class="faint">${whole(entry.positions)}</td>
  </tr>`;
}

function figure(label, value, klass = '') {
  return `<div class="panel figure">
    <span>${esc(label)}</span><b class="mono ${klass}">${value}</b>
  </div>`;
}

/** Where the best outside system sits in the field, the screen in one number. */
function bestSeat(traders) {
  const seat = traders.find((entry) => entry.kind === 'seat');
  if (!seat) return ABSENT;
  const at = num(seat.rank);
  return at == null ? ABSENT : `#${count(at)}`;
}

function balance(conserved) {
  if (conserved === true) return ['Yes', ''];
  if (conserved === false) return ['No', 'amber'];
  return [ABSENT, 'faint'];
}

/* Conservation is the caveat on every number under it, so it is stated above
 * the table rather than beneath it. */
function caveat(conserved) {
  if (conserved === false) {
    return `The venue's books do not balance. Cash and positions across every
      account should sum to what the session started with and right now they do
      not, so read the table as an instrument known to be miscalibrated rather
      than as a result.`;
  }
  if (conserved === true) {
    return `Cash and positions across every account sum to what the session
      started with, so nothing here was created or destroyed on the way.
      Returns are measured on starting cash, which means an account seated a
      minute ago is not comparable to one that has run all session.`;
  }
  return `The venue did not say whether its books balance this session. A
    leaderboard read off a market that is not conserving is not evidence of
    anything, so treat this one as unverified.`;
}

export function standings(store) {
  const table = store?.standings;
  const rows = Array.isArray(table?.standings)
    ? table.standings.filter((entry) => entry && typeof entry === 'object')
    : [];

  // Nothing yet and nothing at all are different states, and the payload
  // arrives on the slow timer, so the first paint of this screen is usually
  // the first of them.
  if (!rows.length) {
    return `<div class="view"><div class="onboard">
      <h2>${table ? 'No accounts on the venue' : 'Reading the standings&hellip;'}</h2>
      <p>${table
        ? 'The exchange published no accounts. Nobody has been seated and no resident agent holds one.'
        : 'Every trader here ranked by what they made: the agents the venue ships with, and the outside systems that connect to it through the public API.'}</p>
    </div></div>`;
  }

  // An unfamiliar kind is ranked with the traders rather than dropped. A
  // participant the screen cannot classify is still a participant, and leaving
  // it out would make the field look smaller than it is.
  //
  // The rank on a row is the venue's, taken over the whole population, so
  // lifting the venue's own accounts out of the list can leave a gap in the
  // sequence. Closing it would mean re-ranking, which is a claim this screen
  // is not the one entitled to make.
  const venue = rows.filter((entry) => entry.kind === 'operator');
  const traders = rows.filter((entry) => entry.kind !== 'operator').sort(byStanding);
  const seats = num(table.counts?.seat)
    ?? traders.filter((entry) => entry.kind === 'seat').length;
  const [balanced, balanceClass] = balance(table.conserved);
  const at = num(table.as_of);
  // Amber is emphasis, and there is nothing to emphasise about a figure the
  // venue could not give: an absent value is greyed like every other one.
  const best = bestSeat(traders);

  return `<div class="view">
    <div class="figures" data-region="standings:figures">
      ${figure('Best Outside Seat', best, best === ABSENT ? 'faint' : 'amber')}
      ${figure('Seats Trading', count(seats))}
      ${figure('As Of', at == null ? ABSENT : clock(at * NS_PER_SECOND))}
      ${figure('Books Balance', balanced, balanceClass)}
    </div>

    <p class="note" style="margin-bottom:16px" data-region="standings:caveat">
      ${caveat(table.conserved)}</p>

    <section class="class-block" data-region="standings:traders">
      <div class="class-head">
        <h2>Standings</h2>
        <span>Outside systems and resident agents, ranked together on return over starting cash.</span>
        <b class="mono">${count(traders.length)}</b>
      </div>
      <div class="panel"><div class="panel-body">
        <table>
          <thead><tr>
            <th style="text-align:right">Rank</th>
            <th style="text-align:left">Trader</th>
            <th style="text-align:left">Kind</th>
            <th>Return</th><th>P&amp;L</th><th>Equity</th><th>Positions</th>
          </tr></thead>
          <tbody>${traders.map(traderRow).join('')}</tbody>
        </table>
      </div></div>
    </section>

    ${venue.length ? `<section class="class-block" data-region="standings:venue">
      <div class="class-head">
        <h2>Venue Accounts</h2>
        <span>The session operator, the match operator and the fee treasury. Not traders, so not ranked.</span>
        <b class="mono">${count(venue.length)}</b>
      </div>
      <div class="panel"><div class="panel-body">
        <table>
          <thead><tr>
            <th>Account</th><th>Equity</th><th>P&amp;L</th><th>Positions</th>
          </tr></thead>
          <tbody>${venue.map(venueRow).join('')}</tbody>
        </table>
      </div></div>
    </section>` : ''}
  </div>`;
}

File: dashboard/standings.py
"""How everyone on this exchange is actually doing, ranked.

Backs the leaderboard. The question it answers is the comparative one: not
"what is this agent doing right now", which `MarketRunner.agents` answers a row
per kernel agent, and not "what jobs are being done here", which
`dashboard.participants.directory` answers a row per role. This one puts a
market maker, a person at a browser and an outside system that has never
imported this package on the same table and sorts them by return.

**The roster is `venue.accounts`, not `market.agents`, and that is the whole
point.** `MarketRunner.agents` walks `self.market.agents`, so it can only ever
report participants this market's own kernel constructed. The two systems in
`connectors/` reach the venue over the published HTTP API and hold nothing but
an account, and they are precisely what a leaderboard exists to compare the
resident agents against. Walking the accounts also picks up the venue's own
treasury, which no roster built from agents can see: it has an account and it
has never had an agent.

The trade in the other direction is real and is the right one. An agent seated
without an account does not appear here, because the venue opens accounts
lazily and an account is where every figure on this table comes from. A row
with no cash, no equity and no return is not a competitor, it is an empty line.

**Money is a string in price units.** `Account.equity` answers in the integer
*minor* units the ledger is kept in, and publishing that raw has cost this
project real bugs twice: a maker worth 113,125,513.21 drawn on the participants
table as "113125513.21M", and a person's own seat drawn as "143745.00M" beside
a header reading "143.7k". `dashboard/state.py` records both. Everything here
goes out through `arena.portfolio.money.from_money`, as a string, exactly as
`Account.to_dict` and every other payload in this package does.

**A missing number is `None` and never "0.00".** They are different facts and
the difference is the leaderboard's whole job: `"0.00"` says a book is flat,
`None` says nobody could value it. The same rule covers `return_pct` on an
account that opened at zero, which the venue's treasury does by construction,
and `fills` for an account with no agent behind it, which every API client is.

Nothing here raises on a degenerate market. A venue that cannot produce marks,
an account whose positions cannot be read, a market with no kernel: each one
costs the fields it actually covers and leaves the rest standing. An operator
screen that returns 500 because one agent is unaccounted for is worse than one
showing a gap where that agent's equity should be.
"""

from __future__ import annotations

from decimal import Decimal
from typing import Any

from arena.market.venue import FEE_ACCOUNT_ID
from arena.portfolio.money import from_money
from arena.sim.time import NANOS_PER_SECOND

__all__ = ["standings"]

# The three groups, fixed here so a header can draw a total for each of them
# even on a market where nobody is in one. A missing key and a zero are the
# same fact for a count, unlike for a price.
KINDS = ("resident", "seat", "operator")


# -- classification --------------------------------------------------------
#
# Agent classes are imported inside the functions that test against them, the
# way `state.py` and `participants.py` both do it. A reporting surface has no
# business dragging every agent implementation into the import graph of
# whatever asked it a question.


def _operator_label(agent_id: str, agent: Any) -> str | None:
    """The venue's own machinery, or None if this is a participant.

    Tested first, before anything else, because two of the three are seated in
    `market.agents` alongside the traders. They are there because in a discrete
    event simulation only a participant with a wakeup can decide that time has
    passed, not because they trade, and a leaderboard that filed the session
    operator as a competing agent would be reporting the referee's score.

    The treasury goes by id because it is the one that has no agent at all: it
    is an account and only an account, deliberately, so the conservation check
    counts the fees it holds.
    """
    if agent_id == str(FEE_ACCOUNT_ID):
        return "Venue treasury"
    if agent is None:
        return None
    try:
        from arena.market.match_operator import MatchOperator
        from arena.market.operator import SessionOperator
        from arena.market.venue_agent import VenueAgent

        for kind, label in (
            (SessionOperator, "Session operator"),
            (MatchOperator, "Match operator"),
            (VenueAgent, "Venue"),
        ):
            if isinstance(agent, kind):
                return label
    except Exception:  # pragma: no cover - an agent module that will not import
        return None
    return None


def _resident_label(agent: Any) -> str:
    """What a resident agent does, in words, capitalised for a column.

    The ladder and its order are `dashboard/state.py`'s `_role`, which derives
    the answer by isinstance rather than by class name for a reason worth
    repeating: two tests went looking for the string "MarketMaker" and quietly
    stopped finding the options maker the day it arrived. A specialised maker
    is still a maker, so `SurfaceMarketMaker` and `MatchMaker` both read
    "Market maker" here, and the `id` column is where they are told apart.

    Falls back to the class name rather than to the id, because a reader who
    does not recognise `FlowTrader` still learns more from it than from
    `flow-03`.
    """
    try:
        from arena.agents.arbitrageur import Arbitrageur
        from arena.agents.bayesian import BayesianFundamental
        from arena.agents.flow import FlowTrader
        from arena.agents.fundamental import FundamentalTrader
        from arena.agents.market_maker import MarketMaker
        from arena.agents.noise import NoiseTrader

        for kind, label in (
            (MarketMaker, "Market maker"),
            (Arbitrageur, "Arbitrageur"),
            (BayesianFundamental, "Informed"),
            (FundamentalTrader, "Informed"),
            (FlowTrader, "Order flow"),
            (NoiseTrader, "Uninformed"),
        ):
            if isinstance(agent, kind):
                return label
    except Exception:  # pragma: no cover - an agent this ladder cannot classify
        pass
    return type(agent).__name__


def _seat_label(agent_id: str, seat: Any) -> str:
    """A person's own name, or the id when there is nobody to ask.

    An API client that holds an account without ever having taken a seat has no
    name anywhere in the market, and inventing one would be worse than showing
    the id it authenticates as.
    """
    name = str(getattr(seat, "display_name", "") or "").strip()
    return name or agent_id


def _classify(agent_id: str, agent: Any, seat: Any) -> tuple[str, str]:
    """Which group an account is in, and what to call it.

    **Seat is tested before resident, and the order is load bearing.**
    `LiveMarket.seat` appends the new `HumanAgent` to `market.agents` and joins
    it to the kernel, so a browser trader and an API client are kernel agents
    in every sense that matters to a roster built by walking that list. Asking
    "is this a kernel agent" first would file every person on the exchange as a
    resident bot, which is the one distinction this table exists to draw.

    The fall through is `seat` rather than `resident` for the same reason. An
    account with no agent object and no entry in `market.traders` is what an
    outside client looks like from in here: it reached the venue over the API,
    it holds an account, and nothing in this process built it.
    """
    operator = _operator_label(agent_id, agent)
    if operator is not None:
        return "operator", operator
    if seat is not None:
        return "seat", _seat_label(agent_id, seat)
    if agent is not None:
        return "resident", _resident_label(agent)
    return "seat", agent_id


# -- the figures -----------------------------------------------------------


def _money(minor: Any) -> str | None:
    """Minor units as a price-unit string, or None if they will not convert."""
    try:
        return str(from_money(int(minor)))
    except Exception:  # pragma: no cover - a balance that is not an integer
        return None


def _equity_minor(account: Any, marks: dict[str, Any] | None) -> int | None:
    """Equity in minor units, or None when this account cannot be valued.

    Kept in minor units here because `pnl` and `return_pct` are both derived
    from it, and deriving them from the rounded string would put a float back
    in the middle of a ledger whose entire argument is that it does not have
    one.
    """
    if marks is None:
        return None
    try:
        return int(account.equity(marks))
    except Exception:  # pragma: no cover - a position with no mark to value it
        return None


def _unrealized(account: Any, marks: dict[str, Any] | None) -> str | None:
    if marks is None:
        return None
    try:
        return _money(account.unrealized_pnl(marks))
    except Exception:  # pragma: no cover - a position with no mark to value it
        return None


def _attribute(account: Any, name: str) -> str | None:
    """One money field off the account, as a price-unit string."""
    try:
        return _money(getattr(account, name))
    except Exception:  # pragma: no cover - an account missing a ledger field
        return None


def _return_pct(equity: int | None, starting: int | None) -> float | None:
    """Return on opening capital, in percent.

    None on an account that opened at zero, which is not a degenerate case: the
    venue's treasury opens empty on purpose so its balance equals what the fee
    schedule collected. A percentage of nothing is undefined rather than
    infinite, and a leaderboard is not the place to decide which.

    Computed on the integer minor units and converted once at the end, so the
    ranking is a fact about the ledger rather than about two roundings.
    """
    if equity is None or not starting:
        return None
    try:
        return float(Decimal(equity - starting) / Decimal(starting) * 100)
    except Exception:  # pragma: no cover - a starting balance that will not divide
        return None


def _held(account: Any) -> int | None:
    """Symbols this account has a position in, counted on non-zero quantity.

    Not `len(account.positions)`. `Account.position` creates a flat `Position`
    on first enquiry and leaves it in the dict, so the raw length counts every
    symbol ever asked about rather than every symbol held. Measured on seed 7
    after sixty simulated seconds: 1,127 position entries across the venue's
    24 accounts against 1,092 actually held, with `noise-09` carrying 50
    entries for 43 positions. A trader that had gone home flat would read as
    still holding everything it ever touched.
    """
    try:
        positions = account.positions
    except Exception:  # pragma: no cover - an account with no position book
        return None
    try:
        return sum(1 for p in positions.values() if getattr(p, "quantity", 0))
    except Exception:  # pragma: no cover - a position book that will not walk
        return None


def _fills(agent: Any) -> int | None:
    """Fills this participant has taken, or None where nobody counted.

    `TradingAgent` counts its own, so a resident and a browser seat both have a
    real figure. An account reached over the API has no agent in this process
    and therefore no counter, and the honest answer there is that this market
    does not know rather than zero.
    """
    if agent is None:
        return None
    value = getattr(agent, "fills", None)
    try:
        return None if value is None else int(value)
    except Exception:  # pragma: no cover - a counter that is not a number
        return None


def _row(
    agent_id: str,
    account: Any,
    agent: Any,
    seat: Any,
    marks: dict[str, Any] | None,
    halted: set[str] | None,
) -> dict[str, Any]:
    """One competitor.

    Every field is guarded on its own, because one unreadable account is not a
    reason to lose the other twenty three rows.
    """
    kind, label = _classify(agent_id, agent, seat)
    equity = _equity_minor(account, marks)
    try:
        starting = int(account.starting_cash)
    except Exception:  # pragma: no cover - an account with no opening balance
        starting = None
    return {
        "id": agent_id,
        "label": label,
        "kind": kind,
        "starting_cash": None if starting is None else _money(starting),
        "cash": _attribute(account, "cash"),
        "equity": None if equity is None else _money(equity),
        "pnl": (
            None
            if equity is None or starting is None
            else _money(equity - starting)
        ),
        "return_pct": _return_pct(equity, starting),
        "realized": _attribute(account, "realized_pnl"),
        "unrealized": _unrealized(account, marks),
        "collateral": _attribute(account, "posted_collateral"),
        "free_cash": _attribute(account, "free_cash"),
        "positions": _held(account),
        "fills": _fills(agent),
        # None rather than False when the venue cannot produce its stop list.
        # "Not halted" is a claim about a participant and this would be a fact
        # about the venue, and a leaderboard that quietly reports one as the
        # other is the same defect as a missing price drawn as zero.
        "halted": None if halted is None else agent_id in halted,
    }


def _order(row: dict[str, Any]) -> tuple:
    """Best return first, unrankable last, ties broken on id.

    The tiebreak is what keeps the table still. Without it, two accounts on an
    identical return swap places between polls at Python's discretion and the
    rows move under the cursor of whoever is reading them, which looks exactly
    like the market doing something.
    """
    pct = row["return_pct"]
    return (pct is None, 0.0 if pct is None else -pct, row["id"])


# -- the payload -----------------------------------------------------------


def standings(market: Any) -> dict[str, Any]:
    """Every account on this venue, ranked by return on opening capital.

    Takes a `LiveMarket` and never raises on one.

    `kind` says where a participant came from rather than what it trades.
    **resident** is an agent this market's own kernel constructed and wakes.
    **seat** is a person at a browser or a system calling the published API,
    which are the same thing from in here: an account somebody outside this
    process is trading. **operator** is the venue's own machinery, which is
    ranked alongside everyone else because its numbers are as real as theirs,
    and which will generally sit wherever a non-trader sits.

    `conserved` is the venue's own conservation check reduced to a flag: total
    equity minus total starting capital, which has to be exactly zero because
    trading moves value between participants rather than creating it. It is
    `None` when the check could not be run, since a leaderboard claiming the
    books balance when nothing verified it is worse than one saying so.
    """
    venue = getattr(market, "venue", None)

    accounts: dict[str, Any] = {}
    if venue is not None:
        try:
            accounts = {str(k): v for k, v in venue.accounts.items()}
        except Exception:  # pragma: no cover - a venue that cannot list accounts
            accounts = {}

    # Marks once for the whole table, not once per account. `Venue.marks`
    # prices every listed symbol, and calling it per row would price the
    # listing twenty four times over to draw one screen.
    marks = None
    if venue is not None:
        try:
            marks = venue.marks()
        except Exception:  # pragma: no cover - a book with nothing to mark against
            marks = None

    agents: dict[str, Any] = {}
    for agent in list(getattr(market, "agents", None) or []):
        agent_id = getattr(agent, "agent_id", None)
        if agent_id is not None:
            agents[str(agent_id)] = agent
    # The venue agent is added straight to the kernel rather than to
    # `market.agents`, so it is invisible to the loop above. It is included so
    # that if it ever does hold an account it is classified as the venue's own
    # machinery instead of falling through to the outside-client default.
    venue_agent = getattr(market, "venue_agent", None)
    if venue_agent is not None:
        venue_agent_id = getattr(venue_agent, "agent_id", None)
        if venue_agent_id is not None:
            agents.setdefault(str(venue_agent_id), venue_agent)

    seats: dict[str, Any] = {}
    try:
        seats = {str(k): v for k, v in market.traders.items()}
    except Exception:  # pragma: no cover - a market with no seat register
        seats = {}

    halted: set[str] | None
    try:
        halted = {str(a) for a in venue.halted_participants}
    except Exception:  # pragma: no cover - a venue with no stop list
        halted = None

    rows = [
        _row(agent_id, account, agents.get(agent_id), seats.get(agent_id), marks, halted)
        for agent_id, account in accounts.items()
    ]
    rows.sort(key=_order)
    for position, row in enumerate(rows, 1):
        row["rank"] = position

    counts = dict.fromkeys(KINDS, 0)
    for row in rows:
        counts[row["kind"]] = counts.get(row["kind"], 0) + 1

    as_of: float | None
    try:
        as_of = int(market.kernel.now) / NANOS_PER_SECOND
    except Exception:  # pragma: no cover - a market with no kernel to ask
        as_of = None

    conserved: bool | None
    try:
        conserved = int(venue.conservation_check()) == 0
    except Exception:  # pragma: no cover - a venue that cannot value its accounts
        conserved = None

    return {
        "standings": rows,
        "as_of": as_of,
        "counts": counts,
        "conserved": conserved,
    }
